overlord/snapstate/snapmgr.go

```diff
@@ -21,38 +21,100 @@
 package snapstate
 
 import (
+	"fmt"
+
+	"github.com/ubuntu-core/snappy/i18n"
 	"github.com/ubuntu-core/snappy/overlord/state"
+	"github.com/ubuntu-core/snappy/progress"
+	"github.com/ubuntu-core/snappy/snappy"
 )
 
+// Install initiates a change installing snap.
+// Note that the state must be locked by the caller.
+func Install(change *state.Change, snap, channel string) error {
+	t := change.NewTask("install-snap", fmt.Sprintf(i18n.G("Installing %q"), snap))
+	t.Set("name", snap)
+	t.Set("channel", channel)
+
+	return nil
+}
+
+// Remove initiates a change removing snap.
+// Note that the state must be locked by the caller.
+func Remove(change *state.Change, snap string) error {
+	t := change.NewTask("remove-snap", fmt.Sprintf(i18n.G("Removing %q"), snap))
+	t.Set("name", snap)
+
+	return nil
+}
+
+type backendIF interface {
+	Install(name, channel string, flags snappy.InstallFlags, meter progress.Meter) (string, error)
+	Remove(name string, flags snappy.RemoveFlags, meter progress.Meter) error
+}
+
+type snappyBackend struct {
+}
+
+func (s *snappyBackend) Install(name, channel string, flags snappy.InstallFlags, meter progress.Meter) (string, error) {
+	return snappy.Install(name, channel, flags, meter)
+}
+
+func (s *snappyBackend) Remove(name string, flags snappy.RemoveFlags, meter progress.Meter) error {
+	return snappy.Remove(name, flags, meter)
+}
+
 // SnapManager is responsible for the installation and removal of snaps.
-type SnapManager struct{}
+type SnapManager struct {
+	state   *state.State
+	backend backendIF
+
+	runner *state.TaskRunner
+}
 
 // Manager returns a new snap manager.
 func Manager() (*SnapManager, error) {
 	return &SnapManager{}, nil
 }
 
-// Install initiates a change installing snap.
-func (m *SnapManager) Install(snap string) error {
-	return nil
+func (m *SnapManager) doInstallSnap(t *state.Task) error {
+	var name, channel string
+	t.State().Lock()
+	t.Get("name", &name)
```

Comment:
It is actually unreasonable to assume it always works. The state comes from disk, potentially written by an unknown version of snappy (from the future or from the past).